### PR TITLE
Return coverage collections or coverage

### PR DESCRIPTION
**Pull request by ernstdevreede**
This MR adds code to enable EDR to return either a Coverage or a CoverageCollection.
If all the returned data fits in one coverage, it returns a coverage, else it returns a CoverageCollection.

### CODE REVIEW

File: python/python_fastapi_server/routers/edr.py [hunk 1]
```diff
@@ -328,20 +365,23 @@ async def get_collectioninfo_for_id(
...
     vertical = None
     vertical_dim = get_vertical_dim_for_collection(
-        edr_collectioninfo, wmslayers, edr_collectioninfo["parameters"][0]["name"]
-    )
+        edr_collectioninfo, wmslayers,
+        edr_collectioninfo["parameters"][0]["name"])
     if vertical_dim:
         vertical = Vertical(**vertical_dim)
 
+    print("interval:", interval)
```
Comment (maartenplieger) [inline python/python_fastapi_server/routers/edr.py]: Please remove

File: python/python_fastapi_server/routers/edr.py [hunk 2]
```diff
@@ -492,18 +533,20 @@ def get_time_values_for_range(rng: str) -> list[str]:
...
+        edr_collectioninfo: dict,
+        wmslayers,
+        parameter: str = None) -> tuple[list[list[str]], list[str]]:
     """
     Returns a list of times based on the time dimensions, it does a WMS GetCapabilities to the given dataset (cached)
 
     It does this for given parameter. When the parameter is not given it will do it for the first Layer in the GetCapabilities document.
     """
-    logger.info("get_times_for_dataset(%s,%s)", edr_collectioninfo["name"], parameter)
+    logger.info("get_times_for_dataset(%s,%s)", edr_collectioninfo["name"],
```
Comment (maartenplieger) [inline python/python_fastapi_server/routers/edr.py]: Is this logging needed here?

File: python/python_fastapi_server/routers/edr.py [hunk 3]
```diff
@@ -9,18 +9,21 @@
 """
```
Comment (maartenplieger) [inline python/python_fastapi_server/routers/edr.py]: Can you please use the same linter as before? I think it should be black? Can do this in vscode by selecting the formatter for this project.

File: python/python_fastapi_server/routers/edr.py (hunk 2 above)
Comment (maartenplieger) [inline python/python_fastapi_server/routers/edr.py]: remove?

File: python/python_fastapi_server/routers/edr.py [hunk 4]
```diff
@@ -9,14 +9,18 @@
...
+import functools
+import operator
 import json
 import logging
 import os
 import re
 from datetime import datetime, timezone
+from dateutil.relativedelta import relativedelta
 from typing import Union
+import pickle
```
Comment (maartenplieger) [inline python/python_fastapi_server/routers/edr.py]: Please remove

File: python/python_fastapi_server/routers/edr.py [hunk 5]
```diff
@@ -210,13 +216,13 @@ async def get_point_value(
...
     "/collections/{collection_name}/position",
-    response_model=Coverage,
+    response_model=Coverage | CoverageCollection,
     response_class=CovJSONResponse,
     response_model_exclude_none=True,
 )
 @edrApiApp.get(
     "/collections/{collection_name}/instances/{instance}/position",
-    response_model=Coverage,
+    response_model=Coverage | CoverageCollection,
```
Comment (maartenplieger) [inline python/python_fastapi_server/routers/edr.py]: Are these two API defs correct for the same function?

File: requirements.txt [hunk 6]
```diff
@@ -93,26 +93,28 @@ packaging==24.0
...
     # via
     #   -r requirements.in
     #   owslib
 pyyaml==6.0.1
     # via owslib
-redis==5.0.3
+redis==5.0.4
     # via -r requirements.in
-requests==2.31.0
+requests==2.32.2
```
Comment (mgrunbauer) [inline requirements.txt]: Note: the updated packages from `requirements.txt`, besides the `edr-pydantic` package, are already updated in master. This was done today in: https://github.com/KNMI/adaguc-server/pull/370/files.

I think you can merge master in your branch to solve this (maybe github resolves this automatically? I'm not sure).

**Comment by mgrunbauer** [comment]
For 24 hourly data (e.g. `air_temperature_max_24h_hagl`), the first 24 hours are missing, so the first data value occurs on T=25 (first hour of the second day).

If you query any datetime range within the first 24 hours, the server returns an 500 internal server error.

Adaguc server logs the following:
```
[E:000:pid21: adagucserverEC/CRequest.cpp:3247] 2024-05-23T09:48:51.950Z ADAGUC Server: Parameter I missing
[E:001:pid21: adagucserverEC/CRequest.cpp:3251] 2024-05-23T09:48:51.950Z ADAGUC Server: Parameter J missing
[D:002:pid21: adagucserverEC/CRequest.cpp:570]  2024-05-23T09:48:51.950Z WMS GETFEATUREINFO air_temperature_max_24h_hagl
[E:003:pid21: adagucserverEC/CRequest.cpp:1762] 2024-05-23T09:48:51.960Z Invalid dimensions values: No data available for layer air_temperature_max_24h_hagl
[E:004:pid21: adagucserverEC/CRequest.cpp:3259] 2024-05-23T09:48:51.960Z WMS GetFeatureInfo Request failed
```

**Comment by mgrunbauer** [comment]
Minor issue:

When querying with only the urlparams `coords` and `datetime`, and a missing `parameter-name`, you get the following helpful message:
```json
{
  "detail": [
    {
      "type": "missing",
      "loc": [
        "query",
        "parameter-name"
      ],
      "msg": "Field required",
      "input": null,
      "url": "https://errors.pydantic.dev/2.6/v/missing"
    }
  ]
}
```

However when you pass `parameter-name` but the value is empty, you get an internal server error.

Example url:
```bash
https://localhost/edr/collections/uwcw_ha43_nl_2km_pl/instances/202405210000/position?coords=POINT(4.926%2051.970)&datetime=2024-05-21T00:00Z/2024-05-22T00:00Z&parameter-name"
```

File: python/python_fastapi_server/routers/edr.py (hunk 1 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/edr.py]: Removed

File: python/python_fastapi_server/routers/edr.py (hunk 2 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/edr.py]: No

Comment (ernstdevreede) [reply python/python_fastapi_server/routers/edr.py]: Commented out

File: python/python_fastapi_server/routers/edr.py (hunk 5 above)
Comment (ernstdevreede) [reply python/python_fastapi_server/routers/edr.py]: No. Created an extra method for the second API def

File: requirements.txt (hunk 6 above)
Comment (ernstdevreede) [reply requirements.txt]: Done

**Comment by mgrunbauer** [comment]
Just checking: querying both a correct and incorrectly spelled `parameter-name` results in no data. Is this expected? So:
```
parameter-name=air_temperature_pl     -> 200 with correct data
parameter-name=aaa                    -> 400 no data
parameter-name=air_temperature_pl,aaa -> 400 no data
```

**Comment by mgrunbauer** [comment]
Custom dimension issues, found when querying with an extra member dimension (members 1-5 present in netcdf). Configured like this:

```xml
    <Dimension name="member" units="" default="min" hidden="false">member</Dimension>
```

- Querying without urlparam `member`, or with `member=1` gives a response with valid data. However the `member` is not mentioned in the response. I think this should behave similar to how querying height works.
- Querying non existing `member=6` gives a 500 internal server error.
- Querying multiple `member=1,2`, `member=*` or `member=1/2` all give 500 internal server error. Note: querying `member=1/1` does work and behaves like `member=1`.

**Review by maartenplieger — APPROVED** [review]

**Pull request merged**